File: python/problem_28.py
from typing import List


def justify_text(words: List[str], k: int) -> List[str]:
    lines = []
    i = 0
    while i < len(words):
        line_len = len(words[i])
        j = i + 1
        while j < len(words) and line_len + 1 + len(words[j]) <= k:
            line_len += 1 + len(words[j])
            j += 1
        line_words = words[i:j]
        gaps = j - i - 1
        if j == len(words) or gaps == 0:
            line = ' '.join(line_words)
            lines.append(line + ' ' * (k - len(line)))
        else:
            total_chars = sum(len(w) for w in line_words)
            total_spaces = k - total_chars
            base, extra = divmod(total_spaces, gaps)
            line = ''
            for idx, w in enumerate(line_words):
                line += w
                if idx < gaps:
                    line += ' ' * base + (' ' if idx < extra else '')
            lines.append(line)
        i = j
    return lines


def main():
    print("Problem 28:")
    words = ["the", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog"]
    result = justify_text(words, 16)
    expected = ["the  quick brown", "fox  jumps  over", "the   lazy   dog"]
    assert result == expected, f"got {result}"
    assert all(len(line) == 16 for line in result)
    print("All Tests Passed")

File: rust/problem_28.rs
pub fn justify_text(words: &[&str], k: usize) -> Vec<String> {
    let mut lines: Vec<String> = Vec::new();
    let mut i = 0;
    while i < words.len() {
        let mut line_len = words[i].len();
        let mut j = i + 1;
        while j < words.len() && line_len + 1 + words[j].len() <= k {
            line_len += 1 + words[j].len();
            j += 1;
        }
        let line_words = &words[i..j];
        let gaps = j - i - 1;
        if j == words.len() || gaps == 0 {
            let line = line_words.join(" ");
            let pad = k - line.len();
            lines.push(format!("{}{}", line, " ".repeat(pad)));
        } else {
            let total_chars: usize = line_words.iter().map(|w| w.len()).sum();
            let total_spaces = k - total_chars;
            let base = total_spaces / gaps;
            let extra = total_spaces % gaps;
            let mut line = String::new();
            for (idx, w) in line_words.iter().enumerate() {
                line.push_str(w);
                if idx < gaps {
                    line.push_str(&" ".repeat(base));
                    if idx < extra { line.push(' '); }
                }
            }
            lines.push(line);
        }
        i = j;
    }
    lines
}

pub fn problem28() {
    println!("Problem 28:");
    let words = vec!["the","quick","brown","fox","jumps","over","the","lazy","dog"];
    let result = justify_text(&words, 16);
    let expected = vec!["the  quick brown", "fox  jumps  over", "the   lazy   dog"];
    assert_eq!(result, expected);
    assert!(result.iter().all(|l| l.len() == 16));
    println!("All Tests Passed");
}
